File: apps/api/services/director/analysis/reconcile_topics.py
"""Topic reconciliation for chunked analysis (Phase 12)."""
from __future__ import annotations

from typing import Any

from services.director.analysis.plan_chunks import ChunkPlan, overlap_zone


def _merge_adjacent(topics: list[dict[str, Any]]) -> list[dict[str, Any]]:
    if not topics:
        return []
    sorted_topics = sorted(topics, key=lambda t: float(t.get("start", 0)))
    merged: list[dict[str, Any]] = [dict(sorted_topics[0])]

    for topic in sorted_topics[1:]:
        prev = merged[-1]
        same_label = prev.get("topicLabel") == topic.get("topicLabel")
        adjacent = float(topic.get("start", 0)) <= float(prev.get("end", 0)) + 0.5
        if same_label and adjacent:
            prev["end"] = max(float(prev.get("end", 0)), float(topic.get("end", 0)))
            prev["confidence"] = max(
                float(prev.get("confidence", 0)),
                float(topic.get("confidence", 0)),
            )
        else:
            merged.append(dict(topic))
    return merged


def reconcile_topics(
    chunk_outputs: list[tuple[ChunkPlan, list[dict[str, Any]]]],
) -> list[dict[str, Any]]:
    """Merge topic boundaries from overlapping chunk windows."""
    if not chunk_outputs:
        return []
    if len(chunk_outputs) == 1:
        return _merge_adjacent(list(chunk_outputs[0][1]))

    owned: list[dict[str, Any]] = []
    for chunk, topics in chunk_outputs:
        for topic in topics:
            start = float(topic.get("start", 0))
            end = float(topic.get("end", start))
            mid = (start + end) / 2
            if chunk.core_start <= mid <= chunk.core_end:
                owned.append(topic)

    for i in range(len(chunk_outputs) - 1):
        zone = overlap_zone(chunk_outputs[i][0], chunk_outputs[i + 1][0])
        if not zone:
            continue
        zone_start, zone_end = zone
        left = [
            t
            for t in chunk_outputs[i][1]
            if zone_start <= float(t.get("start", 0)) <= zone_end
        ]
        right = [
            t
            for t in chunk_outputs[i + 1][1]
            if zone_start <= float(t.get("start", 0)) <= zone_end
        ]
        for left_topic in left:
            match = next(
                (
                    r
                    for r in right
                    if abs(float(r.get("start", 0)) - float(left_topic.get("start", 0))) < 1.5
                ),
                None,
            )
            if match:
                owned.append(
                    left_topic
                    if float(left_topic.get("confidence", 0))
                    >= float(match.get("confidence", 0))
                    else match
                )
            else:
                owned.append(left_topic)

    return _merge_adjacent(owned)

File: remotion-service/src/lib/analysis/reconcileTopics.ts
import { type ChunkPlan, overlapZone } from "./planChunks";

export interface TopicShift {
  start: number;
  end: number;
  confidence: number;
  topicLabel: string;
}

function mergeAdjacent(topics: TopicShift[]): TopicShift[] {
  if (topics.length === 0) {
    return [];
  }
  const sorted = [...topics].sort((a, b) => a.start - b.start);
  const merged: TopicShift[] = [sorted[0]];

  for (const topic of sorted.slice(1)) {
    const prev = merged[merged.length - 1];
    const sameLabel = prev.topicLabel === topic.topicLabel;
    const adjacent = topic.start <= prev.end + 0.5;
    if (sameLabel && adjacent) {
      prev.end = Math.max(prev.end, topic.end);
      prev.confidence = Math.max(prev.confidence, topic.confidence);
    } else {
      merged.push({ ...topic });
    }
  }
  return merged;
}

/**
 * Merge topic boundaries from overlapping chunk windows.
 * In overlap zones, keep the boundary with higher confidence.
 */
export function reconcileTopics(
  chunkOutputs: Array<{ chunk: ChunkPlan; topics: TopicShift[] }>,
): TopicShift[] {
  if (chunkOutputs.length === 0) {
    return [];
  }
  if (chunkOutputs.length === 1) {
    return mergeAdjacent([...chunkOutputs[0].topics]);
  }

  const owned: TopicShift[] = [];
  for (const { chunk, topics } of chunkOutputs) {
    for (const topic of topics) {
      const mid = (topic.start + topic.end) / 2;
      if (mid >= chunk.coreStart && mid <= chunk.coreEnd) {
        owned.push(topic);
      }
    }
  }

  for (let i = 0; i < chunkOutputs.length - 1; i += 1) {
    const zone = overlapZone(chunkOutputs[i].chunk, chunkOutputs[i + 1].chunk);
    if (!zone) {
      continue;
    }
    const left = chunkOutputs[i].topics.filter(
      (t) => t.start >= zone.start && t.start <= zone.end,
    );
    const right = chunkOutputs[i + 1].topics.filter(
      (t) => t.start >= zone.start && t.start <= zone.end,
    );
    for (const l of left) {
      const match = right.find((r) => Math.abs(r.start - l.start) < 1.5);
      if (match) {
        owned.push(l.confidence >= match.confidence ? l : match);
      } else {
        owned.push(l);
      }
    }
  }

  return mergeAdjacent(owned);
}
